### ISSUE TITLE
Design to avoid having loaders on the graphql context

### ISSUE THREAD

**Issue by d1manson** [graphql/dataloader#158]
Currently in graphql we are encouraged to create a context for each request hitting the server, with the context providing a fresh instance of the loaders to any resolvers that might need them.

This is nice in terms of security and memory leaks etc, but it can result in code being a little opaque because the resolvers pick up the dataloaders from an opaque context object:

```
// my_resolver.js

someQuery(_, key, {loaders}){
   return loaders.someLoader.load(key);
}
```

It would be nice instead if we could do something like this:
```
// my_resolver.js

import someLoader from './some_loader';
...
someQuery(_, key, ctx){
   return someLoader(ctx).load(key);
}
```

Could this be implemented with a WeakMap inside the loader or maybe by adding uniquely-named state on the context object?  The current api could be built on top of this kind of quasi-stateless concept.

Not only would this make the use of loaders more explicit it would remove the need of creating all your loaders upfront for each request which seems a bit wasteful (though probably not a major perf issue in 99% of cases). 

This might be a bit radical in terms of being an unusual way of dealing with state in JS, but I feel like the code ends up being simpler to understand, well maybe not this specific suggested api but something along these lines.

-----

**Update**: it looks like you can do this as follow (no need to change anything in this repo)....

```javascript
\\ loader_with_context.js
const DataLoader = require("dataloader");
module.exports = function factory(batchFunc, opts) {
  const Constructor = (opts && opts.constructor) || DataLoader;
  const store = new WeakMap();

  return function getLoader(ctx) {
    let loader = store.get(ctx);
    if (!loader) {
      loader = new Constructor(keys => batchFunc(keys, ctx), opts);
      store.set(ctx, loader);
    }
    return loader;
  };
};
```

See the top of the post for example usage from the consumer's point of view. Here is how you define a loader:

```javascript
// some_loader.js
import loaderWithContext from 'loader_with_context';
module.exports = loaderWithContext((keys, ctx) => something, opts); 
```

Basically, the `loaderWithContext` is a factory which returns a function expecting a "context" object. In turn that function returns a normal `DataLoader`.  The "context" object is used as a key in a WeakMap into the cache. You could support other kinds of keys/caches, e.g. a LRU/TTL Cache for user ids.

**Comment by leebyron** [comment]
I like this pattern and support it. WeakMaps actually have very good performance, and this is a good way to lazily create loaders if your schema may have a large number of them and you're concerned about boot performance.

Closing this issue since it's non-actionable, but if you have any suggested follow ups I'd be happy to review PRs.
